presto-main/src/test/java/com/facebook/presto/server/TestServer.java

```diff
@@ -460,6 +463,353 @@ public void testCacheControlHeaderExists()
+                .setHeader(PRESTO_SOURCE, "source")
+                .setHeader(PRESTO_CATALOG, "catalog")
+                .setHeader(PRESTO_SCHEMA, "schema")
+                .build();
+
+        QueryResults queryResults = client.execute(request, createJsonResponseHandler(QUERY_RESULTS_CODEC));
+        assertNotNull(queryResults);
+        assertNotNull(queryResults.getId());
+
+        // The actual error code filtering is tested in unit tests
+        // This integration test just verifies the parameters are accepted
+    }
+
+    @Test
+    public void testCrossClusterRetryWithRetryChainPrevention()
+            throws Exception
+    {
+        // Create two test servers to simulate two clusters
+        TestingPrestoServer backupServer = null;
+        HttpClient backupClient = null;
+        try {
+            backupServer = new TestingPrestoServer();
+            backupClient = new JettyHttpClient();
+
+            System.out.println("Primary: " + server.getBaseUrl());
+            System.out.println("Backup: " + backupServer.getBaseUrl());
+
+            // Step 1: Simulate router POSTing query to backup server with retry header
+            URI backupUri = HttpUriBuilder.uriBuilderFrom(backupServer.getBaseUrl())
+                    .replacePath("/v1/statement")
+                    .build();
+
+            String failQuery = format("SELECT fail(%s, 'Simulated remote task error')", REMOTE_TASK_ERROR.toErrorCode().getCode());
+            Request backupRequest = preparePost()
+                    .setUri(backupUri)
+                    .setBodyGenerator(createStaticBodyGenerator(failQuery, UTF_8))
+                    .setHeader(PRESTO_USER, "user")
+                    .setHeader(PRESTO_SOURCE, "source")
+                    .setHeader(PRESTO_CATALOG, "catalog")
+                    .setHeader(PRESTO_SCHEMA, "schema")
+                    .setHeader(PRESTO_SESSION, "query_retry_limit=1")
+                    .setHeader(PRESTO_RETRY_QUERY, "true")  // Router marks this as a retry query
+                    .build();
+
+            // Router POSTs to backup server and gets the query ID for the retry query
+            QueryResults backupResults = backupClient.execute(backupRequest, createJsonResponseHandler(QUERY_RESULTS_CODEC));
+            assertNotNull(backupResults);
+            String retryQueryId = backupResults.getId();
+
+            // Step 2: Configure retry URL to point to the backup cluster's retry endpoint
+            String retryUrl = backupServer.getBaseUrl() + "/v1/statement/queued/retry/" + retryQueryId;
+
+            // Step 3: Router redirects client to primary server with retry parameters
+            URI uri = HttpUriBuilder.uriBuilderFrom(server.getBaseUrl())
+                    .replacePath("/v1/statement")
+                    .addParameter("retryUrl", urlEncode(retryUrl))
+                    .addParameter("retryExpirationInSeconds", "3600")
+                    .build();
```

Comment:
Nit: We can extract the URI-building logic into a separate method and reuse it everywhere.